vortex-array/src/vtable/mod.rs

```diff
@@ -0,0 +1,116 @@
+//! This module contains the VTable definitions for a Vortex encoding.
+
+mod array;
+mod compute;
+mod decode;
+mod encode;
+mod operations;
+mod serde;
+mod validity;
+mod visitor;
+
+use std::fmt::Debug;
+use std::ops::Deref;
+
+pub use array::*;
+pub use compute::*;
+pub use decode::*;
+pub use encode::*;
+pub use operations::*;
+pub use serde::*;
+pub use validity::*;
+pub use visitor::*;
+
+use crate::{Array, Encoding, EncodingId, EncodingRef, IntoArray};
+
+/// The encoding [`VTable`] encapsulates logic for an Encoding type and associated Array type.
+/// The logic is split across several "VTable" traits to enable easier code organization than
+/// simply lumping everything into a single trait.
+///
+/// Some of these vtables are optional, such as the [`SerdeVTable`], which is only required if
+/// the encoding supports serialization.
+///
+/// From this [`VTable`] trait, we derive implementations for the sealed [`Array`] and [`Encoding`]
+/// traits via the [`crate::ArrayAdapter`] and [`crate::EncodingAdapter`] types respectively.
+///
+/// The functions defined in these vtable traits will typically document their pre- and
+/// post-conditions. The pre-conditions are validated inside the [`Array`] and [`Encoding`]
+/// implementations so do not need to be checked in the vtable implementations (for example, index
+/// out of bounds). Post-conditions are validated after invocation of the vtable function and will
+/// panic if violated.
+pub trait VTable: 'static + Sized + Send + Sync + Debug {
```

Comment:
i'm surprised that dyn compatible isn't preferred here

Reply:
Encoding and Array are the "objects" / "dyn-safe" things here. VTable is purely a collection of behaviours. Hence why we can have associated types, take/return `Self::Array`, and other fun things.